runtime-sdk/modules/evm/src/lib.rs

```diff
@@ -551,6 +566,57 @@ impl<Cfg: Config> Module<Cfg> {
     {
         derive_caller::from_tx_auth_info(ctx.tx_auth_info())
     }
+
+    fn encode_evm_result<C: Context>(
+        ctx: &C,
+        evm_result: Result<Vec<u8>, Error>,
+        tx_metadata: callformat::Metadata,
+    ) -> Result<Vec<u8>, Error> {
+        if !Cfg::CONFIDENTIAL {
+            return evm_result;
+        }
+        let call_result = match evm_result {
+            Ok(exit_value) => module::CallResult::Ok(exit_value.into()),
+            Err(e) => module::CallResult::Failed {
+                module: e.module_name().into(),
+                code: e.code(),
+                message: e.to_string(),
+            },
+        };
+        Ok(cbor::to_vec(callformat::encode_result(
+            ctx,
+            call_result,
+            tx_metadata,
+        )))
+    }
+
+    /// Returns the decrypted call data or `None` if this transaction is simulated in
+    /// a context that may not include a key manager (i.e. SimulateCall but not EstimateGas).
+    fn decode_call_data<C: Context>(
+        ctx: &C,
+        data: Vec<u8>,
+        tx_index: usize,
+        assume_km_reachable: bool,
+    ) -> Result<Option<(Vec<u8>, callformat::Metadata)>, Error> {
+        if !Cfg::CONFIDENTIAL {
+            return Ok(Some((data, callformat::Metadata::Empty)));
+        }
+        let call = cbor::from_slice(&data)
+            .map_err(|_| CoreError::InvalidCallFormat(anyhow::anyhow!("invalid packed call")))?;
```

Comment:
Hm, so this will prevent one to use the non-backwards-compatible confidential mode where the encryption happens at the outer layer (e.g. in the actual transaction) as in that case `data` will already be the raw data that EVM expects. We could solve this by storing the original call format in the transaction context (e.g. in `with_tx`).

Then the above check could be changed to:
```rust
if !Cfg::CONFIDENTIAL || ctx.tx_call_format() != CallFormat::Plain {
```
And this would support both variants (would be good to document these two variants in a comment somewhere in this module).

Reply:
Good point. I hadn't considered that people would submit non-web3 compat txs to an Eth-like runtime.